Update to Python 3.12.x (#5736)

dsotirho-ucsc opened the pull request:
<!--
This is the PR template for regular PRs against `develop`. Edit the URL in your
browser's location bar, appending either `&template=anvilprod-promotion.md`,
`&template=prod-promotion.md`, `&template=anvilprod-hotfix.md`, `&template=prod-
hotfix.md`, `&template=backport.md` or `&template=upgrade.md` to switch the
template.
-->

Connected issues: #5736


## Checklist


### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] On ZenHub, PR is connected to all issues it (partially) resolves
- [x] PR description links to connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] PR title references all connected issues
- [x] For each connected issue, there is at least one commit whose title references that issue

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title


### Author (partiality)

- [x] Added `p` tag to titles of partial commits
- [x] This PR is labeled `partial` <sub>or completely resolves all connected issues</sub>
- [x] This PR partially resolves each of the connected issues <sub>or does not have the `partial` label</sub>


### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or is not chained to another PR</sub>
- [x] The blocking PR is labeled `base` <sub>or this PR is not chained to another PR</sub>
- [x] This PR is labeled `chained` <sub>or is not chained to another PR</sub>


### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or the changes introduced by this PR will not require reindexing of any deployment</sub>
- [x] This PR is labeled `reindex:dev` <sub>or the changes introduced by it will not require reindexing of `dev`</sub>
- [x] This PR is labeled `reindex:anvildev` <sub>or the changes introduced by it will not require reindexing of `anvildev`</sub>
- [x] This PR is labeled `reindex:anvilprod` <sub>or the changes introduced by it will not require reindexing of `anvilprod`</sub>
- [x] This PR is labeled `reindex:prod` <sub>or the changes introduced by it will not require reindexing of `prod`</sub>
- [x] This PR is labeled `reindex:partial` and its description documents the specific reindexing procedure for `dev`, `anvildev`, `anvilprod` and `prod` <sub>or requires a full reindex or carries none of the labels `reindex:dev`, `reindex:anvildev`, `reindex:anvilprod` and `reindex:prod`</sub>
- [x] This PR and its connected issues are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>


### Author (upgrading deployments)

- [x] Ran `make docker_images.json` and committed the resulting changes <sub>or this PR does not modify `azul_docker_images`, or any other variables referenced in the definition of that variable</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading deployments</sub>
- [x] This PR is labeled `upgrade` <sub>or does not require upgrading deployments</sub>
- [x] This PR is labeled `deploy:shared` <sub>or does not modify `docker_images.json`, and does not require deploying the `shared` component for any other reason</sub>
- [x] This PR is labeled `deploy:gitlab` <sub>or does not require deploying the `gitlab` component</sub>
- [x] This PR is labeled `deploy:runner` <sub>or does not require deploying the `runner` image</sub>


### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the none of the stable branches (`anvilprod` and `prod`) have temporary hotfixes for any of the issues connected to this PR</sub>


### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not modify `requirements*.txt`, `common.mk`, `Makefile` and `Dockerfile`</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not modify `requirements*.txt`</sub>
- [x] This PR is labeled `reqs` <sub>or does not modify `requirements*.txt`</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not modify functionality that could affect the IT outcome</sub>


### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] PR is awaiting requested review from system administrator
- [x] PR is assigned to only the system administrator


### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] A comment to this PR details the completed security design review
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved connected issues to *Approved* column
- [x] PR is assigned to only the operator


### Operator (before pushing merge the commit)

- [x] Checked `reindex:…` labels and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Ran `_select dev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select dev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Ran `_select anvildev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Checked the items in the next section <sub>or this PR is labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the system administrator <sub>or this PR is not labeled `deploy:gitlab`</sub>


### System administrator

- [x] Background migrations for `dev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Background migrations for `anvildev.gitlab` are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the operator


### Operator (before pushing merge the commit)

- [x] Ran `_select dev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Ran `_select anvildev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Started reindex in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] The title of the merge commit starts with the title of this PR
- [x] Added PR # reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only included `p` if the PR is also labeled `partial`</sub>
- [x] Moved connected issues to *Merged lower* column in ZenHub
- [x] Moved blocked issues to *Triage* <sub>or no issues are blocked on the connected issues</sub>
- [x] Pushed merge commit to GitHub


### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>


### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev`
- [x] Pushed merge commit to GitLab `anvildev`
- [x] Build passes on GitLab `dev`
- [x] Reviewed build logs for anomalies on GitLab `dev`
- [x] Build passes on GitLab `anvildev`
- [x] Reviewed build logs for anomalies on GitLab `anvildev`
- [x] Ran `_select dev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`


### Operator (reindex)

- [x] Deindexed all unreferenced catalogs in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed all unreferenced catalogs in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Deindexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Indexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Indexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Emptied fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Emptied fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>


### Operator

- [x] Propagated the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels to the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] Propagated any specific instructions related to the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels, from the description of this PR to that of the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] PR is assigned to no one


## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

dsotirho-ucsc committed:
[u] Update to Python 3.12.x (#5736)

dsotirho-ucsc committed:
Ignore DeprecationWarning from datetime, add FIXME (#5953)

dsotirho-ucsc committed:
Make `requirements_update` download binaries, not source packages

dsotirho-ucsc committed:
Reformat ternary operation

Resolves finding during `make pep8`

E122 continuation line missing indentation or outdented

dsotirho-ucsc committed:
Fix broken assert of Mock object's `not_called` property

Resolves error during test:

AttributeError: 'not_called' is not a valid assertion. Use a spec for the mock if 'not_called' is meant to be an attribute.

dsotirho-ucsc committed:
[R] Update azul-chalice to 1.31.2+13

For Python 3.12 support

dsotirho-ucsc committed:
[R] Add setuptools as a requirement

Fulfills transitive requirement of distutils, which Python 3.12 removed from the standard library

dsotirho-ucsc committed:
Implement referencing.Registry in JSON schema validation (#6666)

Resolves DeprecationWarning during schema validation:

Traceback (most recent call last):
  File "/Users/daniel/.pyenv/versions/3.12.6/lib/python3.12/unittest/suite.py", line 306, in _tearDownPreviousClass
    tearDownClass()
AssertionError: ["{message : DeprecationWarning('Automatically retrieving remote references can be a security vulnerability and is discouraged by the JSON Schema specifications. Relying on this behavior is deprecated and will shortly become an error. If you are sure you want to remotely retrieve your reference and that it is safe to do so, you can find instructions for doing so via referencing.Registry in the referencing documentation (https://referencing.readthedocs.org).'), category : 'DeprecationWarning', filename : '/Users/daniel/repo/azul1/.venv/lib/python3.12/site-packages/jsonschema/exceptions.py', lineno : 475, line : None}", ...]

dsotirho-ucsc commented:
[5951_IT_2024-09-20.txt](https://github.com/user-attachments/files/17080538/5951_IT_2024-09-20.txt)

dsotirho-ucsc commented:
[5951_IT_2024-09-23.txt](https://github.com/user-attachments/files/17103123/5951_IT_2024-09-23.txt)

Successful manifest generation:
```
$ curl -X 'PUT' \
∙   'https://service.daniel.dev.singlecell.gi.ucsc.edu/fetch/manifest/files?catalog=dcp3&filters=%7B%22projectId%22%3A%7B%22is%22%3A%5B%22455b46e6-d8ea-4611-861e-de720a562ada%22%2C%22c893cb57-5c9f-4f26-9312-21b85be84313%22%5D%7D%7D&format=compact'
{"Status":301,"Location":"https://service.daniel.dev.singlecell.gi.ucsc.edu/fetch/manifest/files/k8Qg1YANPZqETJUYbBfBdyHS7h3RVKhWrtEqC4yr1fEaKGgAAQ==","Retry-After":1}

$ curl 'https://service.daniel.dev.singlecell.gi.ucsc.edu/fetch/manifest/files/k8Qg1YANPZqETJUYbBfBdyHS7h3RVKhWrtEqC4yr1fEaKGgAAQ=='
{"Status":302,"Location":"https://s3.amazonaws.com/edu-ucsc-gi-platform-hca-dev-storage-daniel.us-east-1/manifests/c07839fa-e994-5911-9d2d-0b6d8ab77c0e.6471d936-5eac-53e8-8306-b295d0c977e2.tsv?response-content-disposition=attachment%3Bfilename%3D%22hca-manifest-c07839fa-e994-5911-9d2d-0b6d8ab77c0e.6471d936-5eac-53e8-8306-b295d0c977e2.tsv%22&AWSAccessKeyId=ASIARZFZ7W77W5523TML&Signature=PwFB4EL3FAiz7Nao9uoWhExhfqk%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEIz%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJIMEYCIQD6D48nocv%2BpBHqOL4iucVYcQ89IwOgk4fEi9BpEHnIgwIhAOtjLlB3tNK4qccMK8QhByWeJRd4ms%2F%2B%2BJvPAxK53DzMKpYDCMX%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQBBoMMTIyNzk2NjE5Nzc1IgziiqcZ%2B9tSIIZ9K%2Bcq6gJVE2T%2FmF3khLLgfwbFsZMyzN3dmhE%2B2RL6m5CiSBtxCpmjEDF2%2FUJwAsLhS9rC3QU%2FYhE4SwelFcJDRCoWtZe870DFzZGepsrjbXfWjFmEPAtkbTjNc9IuuYsuTqVhS5p0JCcQ%2BnRv0su6y9zaLa3Q1XmbOC7KcLBogjGx8yMGiVcuy9kN2lXuVajsj94O3WhV7SlvJpwY7O87B9CmkUsGaHRDT%2Br1iEfe0hKeOkF0wK%2B71c7NcH88pw45oyBdgjHB1zGh1xtj7kdbggd7NrpzfXJj7%2BV89xQ8aBO6tlLQ2IoKNVMNKrFgsbIx3gty%2B8Tcst11bKVSLJVd71%2FkC59NGfpq5fr0mTr2%2BAwxKeULS2Y07MGEHRi4333CEru55Yc1B%2BOA%2B%2BYyl4pY6mIMD3E8AX%2B%2FxlsYPPoO5j6mJcVCmwlFajq3vQsKanHFQ7eQsWYSHIeP0fzU8lrhSYDYxro1%2F%2B0KUwmPvoEjDDDp%2F8a3BjqcAaswOVbnwpQBMONsJA3leiwYL5wKH%2BaKfV4ooVTndI5DM%2BuawmVnRD%2BmrM6FN%2FZeG7JLfZauY21l5IfbnIk%2F9y0wSKj716lwcMNLX3XnIp6Lkre1%2BZ6jYMmzGpVHizilZx03cZfQQe15lRL%2FWdYNmE%2FT13DqsmLcfDz0VtpxnpiUGMn%2BygYZglr%2BD4CvSpTdYqz%2BPlWYsfgCzsIAcg%3D%3D&Expires=1727122982","CommandLine":{"cmd.exe":"curl.exe --location --fail --output \"hca-manifest-c07839fa-e994-5911-9d2d-0b6d8ab77c0e.6471d936-5eac-53e8-8306-b295d0c977e2.tsv\" \"https://s3.amazonaws.com/edu-ucsc-gi-platform-hca-dev-storage-daniel.us-east-1/manifests/c07839fa-e994-5911-9d2d-0b6d8ab77c0e.6471d936-5eac-53e8-8306-b295d0c977e2.tsv?response-content-disposition=attachment%3Bfilename%3D%22hca-manifest-c07839fa-e994-5911-9d2d-0b6d8ab77c0e.6471d936-5eac-53e8-8306-b295d0c977e2.tsv%22&AWSAccessKeyId=ASIARZFZ7W77W5523TML&Signature=PwFB4EL3FAiz7Nao9uoWhExhfqk%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEIz%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJIMEYCIQD6D48nocv%2BpBHqOL4iucVYcQ89IwOgk4fEi9BpEHnIgwIhAOtjLlB3tNK4qccMK8QhByWeJRd4ms%2F%2B%2BJvPAxK53DzMKpYDCMX%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQBBoMMTIyNzk2NjE5Nzc1IgziiqcZ%2B9tSIIZ9K%2Bcq6gJVE2T%2FmF3khLLgfwbFsZMyzN3dmhE%2B2RL6m5CiSBtxCpmjEDF2%2FUJwAsLhS9rC3QU%2FYhE4SwelFcJDRCoWtZe870DFzZGepsrjbXfWjFmEPAtkbTjNc9IuuYsuTqVhS5p0JCcQ%2BnRv0su6y9zaLa3Q1XmbOC7KcLBogjGx8yMGiVcuy9kN2lXuVajsj94O3WhV7SlvJpwY7O87B9CmkUsGaHRDT%2Br1iEfe0hKeOkF0wK%2B71c7NcH88pw45oyBdgjHB1zGh1xtj7kdbggd7NrpzfXJj7%2BV89xQ8aBO6tlLQ2IoKNVMNKrFgsbIx3gty%2B8Tcst11bKVSLJVd71%2FkC59NGfpq5fr0mTr2%2BAwxKeULS2Y07MGEHRi4333CEru55Yc1B%2BOA%2B%2BYyl4pY6mIMD3E8AX%2B%2FxlsYPPoO5j6mJcVCmwlFajq3vQsKanHFQ7eQsWYSHIeP0fzU8lrhSYDYxro1%2F%2B0KUwmPvoEjDDDp%2F8a3BjqcAaswOVbnwpQBMONsJA3leiwYL5wKH%2BaKfV4ooVTndI5DM%2BuawmVnRD%2BmrM6FN%2FZeG7JLfZauY21l5IfbnIk%2F9y0wSKj716lwcMNLX3XnIp6Lkre1%2BZ6jYMmzGpVHizilZx03cZfQQe15lRL%2FWdYNmE%2FT13DqsmLcfDz0VtpxnpiUGMn%2BygYZglr%2BD4CvSpTdYqz%2BPlWYsfgCzsIAcg%3D%3D&Expires=1727122982\"","bash":"curl --location --fail --output hca-manifest-c07839fa-e994-5911-9d2d-0b6d8ab77c0e.6471d936-5eac-53e8-8306-b295d0c977e2.tsv 'https://s3.amazonaws.com/edu-ucsc-gi-platform-hca-dev-storage-daniel.us-east-1/manifests/c07839fa-e994-5911-9d2d-0b6d8ab77c0e.6471d936-5eac-53e8-8306-b295d0c977e2.tsv?response-content-disposition=attachment%3Bfilename%3D%22hca-manifest-c07839fa-e994-5911-9d2d-0b6d8ab77c0e.6471d936-5eac-53e8-8306-b295d0c977e2.tsv%22&AWSAccessKeyId=ASIARZFZ7W77W5523TML&Signature=PwFB4EL3FAiz7Nao9uoWhExhfqk%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEIz%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJIMEYCIQD6D48nocv%2BpBHqOL4iucVYcQ89IwOgk4fEi9BpEHnIgwIhAOtjLlB3tNK4qccMK8QhByWeJRd4ms%2F%2B%2BJvPAxK53DzMKpYDCMX%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQBBoMMTIyNzk2NjE5Nzc1IgziiqcZ%2B9tSIIZ9K%2Bcq6gJVE2T%2FmF3khLLgfwbFsZMyzN3dmhE%2B2RL6m5CiSBtxCpmjEDF2%2FUJwAsLhS9rC3QU%2FYhE4SwelFcJDRCoWtZe870DFzZGepsrjbXfWjFmEPAtkbTjNc9IuuYsuTqVhS5p0JCcQ%2BnRv0su6y9zaLa3Q1XmbOC7KcLBogjGx8yMGiVcuy9kN2lXuVajsj94O3WhV7SlvJpwY7O87B9CmkUsGaHRDT%2Br1iEfe0hKeOkF0wK%2B71c7NcH88pw45oyBdgjHB1zGh1xtj7kdbggd7NrpzfXJj7%2BV89xQ8aBO6tlLQ2IoKNVMNKrFgsbIx3gty%2B8Tcst11bKVSLJVd71%2FkC59NGfpq5fr0mTr2%2BAwxKeULS2Y07MGEHRi4333CEru55Yc1B%2BOA%2B%2BYyl4pY6mIMD3E8AX%2B%2FxlsYPPoO5j6mJcVCmwlFajq3vQsKanHFQ7eQsWYSHIeP0fzU8lrhSYDYxro1%2F%2B0KUwmPvoEjDDDp%2F8a3BjqcAaswOVbnwpQBMONsJA3leiwYL5wKH%2BaKfV4ooVTndI5DM%2BuawmVnRD%2BmrM6FN%2FZeG7JLfZauY21l5IfbnIk%2F9y0wSKj716lwcMNLX3XnIp6Lkre1%2BZ6jYMmzGpVHizilZx03cZfQQe15lRL%2FWdYNmE%2FT13DqsmLcfDz0VtpxnpiUGMn%2BygYZglr%2BD4CvSpTdYqz%2BPlWYsfgCzsIAcg%3D%3D&Expires=1727122982'"}}

$ curl -o manifest.tsv 'https://s3.amazonaws.com/edu-ucsc-gi-platform-hca-dev-storage-daniel.us-east-1/manifests/c07839fa-e994-5911-9d2d-0b6d8ab77c0e.6471d936-5eac-53e8-8306-b295d0c977e2.tsv?response-content-disposition=attachment%3Bfilename%3D%22hca-manifest-c07839fa-e994-5911-9d2d-0b6d8ab77c0e.6471d936-5eac-53e8-8306-b295d0c977e2.tsv%22&AWSAccessKeyId=ASIARZFZ7W77W5523TML&Signature=PwFB4EL3FAiz7Nao9uoWhExhfqk%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEIz%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJIMEYCIQD6D48nocv%2BpBHqOL4iucVYcQ89IwOgk4fEi9BpEHnIgwIhAOtjLlB3tNK4qccMK8QhByWeJRd4ms%2F%2B%2BJvPAxK53DzMKpYDCMX%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQBBoMMTIyNzk2NjE5Nzc1IgziiqcZ%2B9tSIIZ9K%2Bcq6gJVE2T%2FmF3khLLgfwbFsZMyzN3dmhE%2B2RL6m5CiSBtxCpmjEDF2%2FUJwAsLhS9rC3QU%2FYhE4SwelFcJDRCoWtZe870DFzZGepsrjbXfWjFmEPAtkbTjNc9IuuYsuTqVhS5p0JCcQ%2BnRv0su6y9zaLa3Q1XmbOC7KcLBogjGx8yMGiVcuy9kN2lXuVajsj94O3WhV7SlvJpwY7O87B9CmkUsGaHRDT%2Br1iEfe0hKeOkF0wK%2B71c7NcH88pw45oyBdgjHB1zGh1xtj7kdbggd7NrpzfXJj7%2BV89xQ8aBO6tlLQ2IoKNVMNKrFgsbIx3gty%2B8Tcst11bKVSLJVd71%2FkC59NGfpq5fr0mTr2%2BAwxKeULS2Y07MGEHRi4333CEru55Yc1B%2BOA%2B%2BYyl4pY6mIMD3E8AX%2B%2FxlsYPPoO5j6mJcVCmwlFajq3vQsKanHFQ7eQsWYSHIeP0fzU8lrhSYDYxro1%2F%2B0KUwmPvoEjDDDp%2F8a3BjqcAaswOVbnwpQBMONsJA3leiwYL5wKH%2BaKfV4ooVTndI5DM%2BuawmVnRD%2BmrM6FN%2FZeG7JLfZauY21l5IfbnIk%2F9y0wSKj716lwcMNLX3XnIp6Lkre1%2BZ6jYMmzGpVHizilZx03cZfQQe15lRL%2FWdYNmE%2FT13DqsmLcfDz0VtpxnpiUGMn%2BygYZglr%2BD4CvSpTdYqz%2BPlWYsfgCzsIAcg%3D%3D&Expires=1727122982'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  330k  100  330k    0     0   456k      0 --:--:-- --:--:-- --:--:--  456k

$ wc -l manifest.tsv
     169 manifest.tsv

$ head -n 3 manifest.tsv
source_id source_spec bundle_uuid bundle_version  file_document_id  file_type file_name file_format read_index  file_size file_uuid file_version  file_crc32c file_sha256 file_content_type file_drs_uri  file_url  cell_suspension.provenance.document_id  cell_suspension.biomaterial_core.biomaterial_id cell_suspension.estimated_cell_count  cell_suspension.selected_cell_type  sequencing_process.provenance.document_id sequencing_protocol.instrument_manufacturer_model sequencing_protocol.paired_end  library_preparation_protocol.library_construction_approach  library_preparation_protocol.nucleic_acid_source  project.provenance.document_id  project.contributors.institution  project.contributors.laboratory project.project_core.project_short_name project.project_core.project_title  project.estimated_cell_count  specimen_from_organism.provenance.document_id specimen_from_organism.diseases specimen_from_organism.organ  specimen_from_organism.organ_part specimen_from_organism.preservation_storage.preservation_method donor_organism.sex  donor_organism.biomaterial_core.biomaterial_id  donor_organism.provenance.document_id donor_organism.genus_species  donor_organism.development_stage  donor_organism.diseases donor_organism.organism_age cell_line.provenance.document_id  cell_line.biomaterial_core.biomaterial_id organoid.provenance.document_id organoid.biomaterial_core.biomaterial_id  organoid.model_organ  organoid.model_organ_part _entity_type  sample.provenance.document_id sample.biomaterial_core.biomaterial_id  sequencing_input.provenance.document_id sequencing_input.biomaterial_core.biomaterial_id  sequencing_input_type
403ef1b1-7b7f-4e13-bf4a-d898f047310a  tdr:bigquery:gcp:datarepo-dev-71926fdc:hca_dev_c893cb575c9f4f26931221b85be84313__20210901_20210903:/0 190f7516-53fe-4a12-82aa-1d5fa7d13e57  2021-04-06T17:53:52.841000Z 010685d2-1eae-4e0f-b898-4e8c91a31010  sequence_file B2_bamtofastq_S1_L001_R1_003.fastq.gz fastq.gz  read1 4624044177  ec65979d-a2a3-4c14-89a3-79fea8915531  2021-02-03T11:13:49.469000Z b74d886d  957c92a323ace169f73057f55e2dee01e62c204364d22c27c177d971a80e97e1  application/gzip; dcp-type=data; dcp-type=data  drs://jade.datarepo-dev.broadinstitute.org/v1_403ef1b1-7b7f-4e13-bf4a-d898f047310a_9416f225-0918-479a-85f5-65c14653a27b https://service.daniel.dev.singlecell.gi.ucsc.edu/repository/files/ec65979d-a2a3-4c14-89a3-79fea8915531?catalog=dcp3&version=2021-02-03T11%3A13%3A49.469000Z  c73a2b4d-9aa7-4cb2-b0b8-3822921bd0fc  SRX4291262  8000  epithelial cell 190f7516-53fe-4a12-82aa-1d5fa7d13e57  Illumina NovaSeq 6000 False 10X 3' v2 sequencing  single cell c893cb57-5c9f-4f26-9312-21b85be84313  EMBL-EBI || John Radcliffe Hospital, University of Oxford || MRC Weatherall Institute of Molecular Medicine || MRC Weatherall Institute of Molecular Medicine (WIMM) || Sir William Dunn School of Pathology, University of Oxford || St Mary's Hospital, Imperial College || Target Discovery Institute  Centre for Pathology || MRC WIMM Centre For Computational Biology || Medical Research Council (MRC) Human Immunology Unit || Nuffield Department of Medicine || Nuffield Department of Surgical Sciences || Wolfson Imaging Centre Oxford EpithelialDiversityHealthInflammation Colonic single epithelial cell census reveals goblet cell drivers of barrier breakdown in inflammatory bowel disease.   fd9980b6-5e71-4510-af31-12a92acf5d90  normal  digestive system  colon   unknown ulcerative_colitis_patient_2  99b5fa9b-9a72-434f-9925-11ec6ec5b7e3  Homo sapiens  human adult stage ulcerative colitis (disease)  36-80 year              specimens fd9980b6-5e71-4510-af31-12a92acf5d90  SAMN09479597  c73a2b4d-9aa7-4cb2-b0b8-3822921bd0fc  SRX4291262  cell_suspension
403ef1b1-7b7f-4e13-bf4a-d898f047310a  tdr:bigquery:gcp:datarepo-dev-71926fdc:hca_dev_c893cb575c9f4f26931221b85be84313__20210901_20210903:/0 3ed5f729-0cba-4f75-b915-f7df6b3fdd9a  2021-04-06T17:54:02.375000Z 01c48ac3-ecb3-4d05-819b-56a1b7f3d138  sequence_file B3_bamtofastq_S1_L001_I1_004.fastq.gz fastq.gz  index1  761022671 8539ca7b-e82b-4406-a757-8cfa914ab8c2  2021-02-03T11:13:49.042000Z f4a72c57  59ed7ff27d5dca3bb48e5558b8264988a238cfaf11955703665de022fe54a550  application/gzip; dcp-type=data; dcp-type=data  drs://jade.datarepo-dev.broadinstitute.org/v1_403ef1b1-7b7f-4e13-bf4a-d898f047310a_c896a85a-a822-46c6-9717-6d7504a96d62 https://service.daniel.dev.singlecell.gi.ucsc.edu/repository/files/8539ca7b-e82b-4406-a757-8cfa914ab8c2?catalog=dcp3&version=2021-02-03T11%3A13%3A49.042000Z  4941a8af-830c-432c-a161-a372d191317e  SRX4291261  8000  epithelial cell 3ed5f729-0cba-4f75-b915-f7df6b3fdd9a  Illumina NovaSeq 6000 False 10X 3' v2 sequencing  single cell c893cb57-5c9f-4f26-9312-21b85be84313  EMBL-EBI || John Radcliffe Hospital, University of Oxford || MRC Weatherall Institute of Molecular Medicine || MRC Weatherall Institute of Molecular Medicine (WIMM) || Sir William Dunn School of Pathology, University of Oxford || St Mary's Hospital, Imperial College || Target Discovery Institute  Centre for Pathology || MRC WIMM Centre For Computational Biology || Medical Research Council (MRC) Human Immunology Unit || Nuffield Department of Medicine || Nuffield Department of Surgical Sciences || Wolfson Imaging Centre Oxford EpithelialDiversityHealthInflammation Colonic single epithelial cell census reveals goblet cell drivers of barrier breakdown in inflammatory bowel disease.   8724c66c-76b5-42a1-8dd2-c89a9c851d11  ulcerative colitis (disease)  digestive system  colon   unknown ulcerative_colitis_patient_2  99b5fa9b-9a72-434f-9925-11ec6ec5b7e3  Homo sapiens  human adult stage ulcerative colitis (disease)  36-80 year              specimens 8724c66c-76b5-42a1-8dd2-c89a9c851d11  SAMN09479598  4941a8af-830c-432c-a161-a372d191317e  SRX4291261  cell_suspension
```

achave11-ucsc commented on UPGRADING.rst:
```diff
@@ -20,6 +20,17 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
+Everyone
+~~~~~~~~
+
+Update Python on your developer machines to version 3.12.6. In your working
+copy, run ``make virtualenv`` and ``make requirements envhook``. Update Google
+Cloud CLI on your developer machines to the latest version.
```
Consider omitting the details, IIRC, Hannes has flagged this type of additions as unnecessary since we (the developers) are familiar with the process. However, if you want to include it, you should mention to follow the instructions in the output of the `make virtualenv` output and proceed with the other steps.

But, you should make sure this change requires upgrading instructions, in the past, updates to Python (a patch version) omitted upgrading instructions because of the checks informing us when we're not running on the appropriate version.

Is there a reason why the Google Cloud CLI should be updated with the Python Upgrade change?

```diff
diff --git a/UPGRADING.rst b/UPGRADING.rst
index b5f1ab7f..a071144d 100644
--- a/UPGRADING.rst
+++ b/UPGRADING.rst
@@ -26,9 +26,9 @@ have too many entries in this file.
 Everyone
 ~~~~~~~~
 
-Update Python on your developer machines to version 3.12.6. In your working
-copy, run ``make virtualenv`` and ``make requirements envhook``. Update Google
-Cloud CLI on your developer machines to the latest version.
+Update Python on your developer machines to version 3.12.6.
+
+Update Google Cloud CLI on your developer machines to the latest version.
 
 
 #6570 Upgrade dependencies 2024-09-16

```

dsotirho-ucsc replied:
> Consider omitting the details, IIRC, Hannes has flagged this type of additions as unnecessary since we (the developers) are familiar with the process.

I've removed the upgrading instructions.

> Is there a reason why the Google Cloud CLI should be updated with the Python Upgrade change?

Unfortunately I no longer have the error message that I encountered. however it should be obvious to a developer if an update is needed on their system. I've removed this from the upgrading instructions.

achave11-ucsc commented on requirements.dev.txt:
```diff
@@ -13,7 +13,7 @@ jq==1.5.0
 moto[s3,sqs,sns,dynamodb,iam]==4.1.13  # match the extras with the backends listed in AzulUnitTestCase._reset_moto
 boto3-stubs[s3,sqs,lambda,dynamodb,iam,ecr,stepfunctions,kms]==1.28.63  # match this with the version of the `boto3` runtime dependency
-openapi-spec-validator==0.5.1
+openapi-spec-validator==0.7.1
```
Equivalent change in `requirements.all.txt` is missing in this commit.

achave11-ucsc replied:
Resolved at PL.

achave11-ucsc commented on requirements.pip.txt:
```diff
@@ -1,3 +1,3 @@
 pip==23.3.2
-setuptools==70.3.0
+setuptools==70.3.0  # keep consistent with requirements.txt
```
Equivalent change in `requirements.all.txt` is missing in this commit.

achave11-ucsc replied:
Resolved at PL.

achave11-ucsc commented on requirements.dev.txt:
```diff
@@ -9,7 +9,7 @@ gitpython==3.1.41
 google-cloud-storage==2.12.0
 google-cloud-bigquery-reservation==1.11.3
-jq==1.3.0
+jq==1.5.0
```
Equivalent change in `requirements.all.txt` is missing in this commit.

achave11-ucsc replied:
Resolved at PL.

achave11-ucsc commented on requirements.dev.txt:
```diff
@@ -3,7 +3,7 @@ awscli==1.29.63
 coverage==6.5.0
 docker==7.1.0
-flake8==5.0.4
+flake8==6.1.0
```
Equivalent change in `requirements.all.txt` is missing in this commit.

achave11-ucsc replied:
Resolved at PL.

achave11-ucsc commented on requirements.txt:
```diff
@@ -7,7 +7,7 @@ chevron==0.14.0
 elasticsearch-dsl==7.4.0
 elasticsearch==7.17.9
-fastavro==1.8.2
+fastavro==1.8.4
```
Equivalent change in `requirements.all.txt` is missing in this commit.

achave11-ucsc reviewed (changes requested):
Looks good, some minor structural consistency issues.

achave11-ucsc commented:
Took the liberty of removing the automatic sprint assignments.

dsotirho-ucsc commented:
[5951_IT_2024-09-24.txt](https://github.com/user-attachments/files/17122258/5951_IT_2024-09-24.txt)

achave11-ucsc reviewed (dismissed):
Approved ✅

hannes-ucsc commented on docker_images.json:
```diff
@@ -15,18 +15,18 @@
-    "docker.io/library/python:3.11.10-slim-bullseye": {
-        "digest": "sha256:21e2cc29feabc9fa1335feb5efd98ebaf323a5963b8670a2212289b1db74cf73",
-        "mirror_digest": "sha256:ccb0c6383303002a6d15d2f52a24eb240f46a16f38984b04012a46c6a3b81560",
+    "docker.io/library/python:3.12.6-slim-bullseye": {
```
Looks like you forgot to address https://github.com/DataBiosphere/azul/issues/5736#issuecomment-2354398118 which is a pretty significant unforced error, since I specifically remember triaging that comment with you to make sure you're aware of it.

dsotirho-ucsc commented:
[5951_IT_2024-09-25.txt](https://github.com/user-attachments/files/17139899/5951_IT_2024-09-25.txt)

dsotirho-ucsc commented:
[5951_IT_2024-09-25.txt](https://github.com/user-attachments/files/17140293/5951_IT_2024-09-25.txt)

achave11-ucsc commented on README.md:
````diff
@@ -2319,10 +2319,10 @@ If you add a dependency on a package with native code, you need to build the
 
 ```
-(.venv) ~/workspace/hca/azul$ docker run -it -v ${project_root}/:/root/azul python:3.11.6-bullseye bash
+(.venv) ~/workspace/hca/azul$ docker run -it -v ${project_root}/:/root/azul python:3.12.6-slim-bullseye bash
````
https://github.com/DataBiosphere/azul/issues/5736#issuecomment-2354398118

dsotirho-ucsc commented:
[5951_IT_2024-09-27.txt](https://github.com/user-attachments/files/17171988/5951_IT_2024-09-27.txt)

hannes-ucsc commented on requirements.dev.txt:
```diff
@@ -3,7 +3,7 @@ awscli==1.29.63
 coverage==6.5.0
 docker==7.1.0
-flake8==5.0.4
+flake8==7.1.1
```
Change commit message body from 

Resolves errors originating from flake8 failing to parse f-strings

to 

Resolves errors originating from flake8 failing to parse f-strings under Python 3.12

hannes-ucsc commented on terraform/gitlab/gitlab.tf.json.template.py:
```diff
@@ -1077,7 +1077,7 @@ def sq(*words):
                                           if split_tunnel else
                                           all_ipv4
-                                      ],
+                                      ],  # noqa: E122 https://github.com/PyCQA/pycodestyle/issues/372
```
Is this still necessary after the flake8 update? What the relationship between flake8 and pycodestyle?

dsotirho-ucsc replied:
> Is this still necessary after the flake8 update?

Yes, both `make pep8` and PyCharm flag this as E122 without the `noqa` comment.

> What the relationship between flake8 and pycodestyle?

flake8 uses pycodestyle for its detection of `E` prefixed error codes.

From: https://flake8.pycqa.org/en/latest/internal/contributing.html#filing-a-bug
> Error codes starting with E and W should be reported to [pycodestyle](https://github.com/pycqa/pycodestyle).

dsotirho-ucsc commented:
[5951_IT_2024-09-30.txt](https://github.com/user-attachments/files/17197092/5951_IT_2024-09-30.txt)

hannes-ucsc reviewed (changes requested):
After these changes, this can go directly to @nadove-ucsc for final approval.

```diff
Index: requirements.txt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/requirements.txt b/requirements.txt
--- a/requirements.txt	(revision ca7b32f0b110d3b4b1f6144c44c418f3150ecd80)
+++ b/requirements.txt	(date 1727811638913)
@@ -18,7 +18,7 @@
 msgpack==1.0.7
 requests==2.32.2
 rsa==4.7.2  # resolve ambiguity with build-time dependency
-setuptools==70.3.0  # Keep consistent with requirements.pip.txt. Python 3.12 removed distutils, we depend on it transitively, setuptools includes a vendored copy
+setuptools==70.3.0  # Keep consistent with requirements.pip.txt. Python 3.12 removed distutils, which we depended on transitively. Luckily, setuptools includes a vendored copy.
 urllib3==1.26.19
 werkzeug==3.0.3
 -r requirements.trans.txt
Index: terraform/gitlab/gitlab.tf.json.template.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/terraform/gitlab/gitlab.tf.json.template.py b/terraform/gitlab/gitlab.tf.json.template.py
--- a/terraform/gitlab/gitlab.tf.json.template.py	(revision ca7b32f0b110d3b4b1f6144c44c418f3150ecd80)
+++ b/terraform/gitlab/gitlab.tf.json.template.py	(date 1727811162273)
@@ -1073,11 +1073,7 @@
                 'egress': [
                     vpc.security_rule(description='Any traffic to the '
                                                   f"{'VPC' if split_tunnel else 'internet'}",
-                                      cidr_blocks=[
-                                          '${aws_vpc.gitlab.cidr_block}'
-                                          if split_tunnel else
-                                          all_ipv4
-                                      ],  # noqa: E122 https://github.com/PyCQA/pycodestyle/issues/372
+                                      cidr_blocks=['${aws_vpc.gitlab.cidr_block}' if split_tunnel else all_ipv4],
                                       protocol=-1,
                                       from_port=0,
                                       to_port=0),
```

dsotirho-ucsc commented:
[5951_IT_2024-10-02.txt](https://github.com/user-attachments/files/17234475/5951_IT_2024-10-02.txt)

dsotirho-ucsc commented:
Build [failure](https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/101434) due to combination of FIPS mode & the new bookworm-based Python image can be replicated with these commands:

```
$ sudo /usr/bin/docker run -it --rm --network gitlab-runner-net --env DOCKER_HOST=tcp://gitlab-dind:2375 --dns 172.21.0.2 --dns 169.254.169.253 122796619775.dkr.ecr.us-east-1.amazonaws.com/docker.io/library/docker:27.2.1 \
docker run -it --volume /var/run/docker.sock:/var/run/docker.sock docker.gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/runner:latest \
docker run -it docker.io/library/python:3.12.6-slim-bookworm \
bash

# apt-get update && apt-get upgrade -y && apt-get -y install build-essential curl unzip

# install -m 0755 -d /etc/apt/keyrings

# curl -fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc

# echo "deb [arch="$(dpkg --print-architecture)" signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/debian "$(. /etc/os-release && echo "$VERSION_CODENAME")" stable" | tee /etc/apt/sources.list.d/docker.list

# apt-get update
```

Terminal log:
```
$ ssh gitlab.hcadev

[ec2-user@ip-172-21-0-99 ~]$ sudo /usr/bin/docker run -it --rm --network gitlab-runner-net --env DOCKER_HOST=tcp://gitlab-dind:2375 --dns 172.21.0.2 --dns 169.254.169.253 122796619775.dkr.ecr.us-east-1.amazonaws.com/docker.io/library/docker:27.2.1 \
> docker run -it --volume /var/run/docker.sock:/var/run/docker.sock docker.gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/runner:latest \
> docker run -it docker.io/library/python:3.12.6-slim-bookworm \
> bash

root@02a62aa013df:/# apt-get update && apt-get upgrade -y && apt-get -y install build-essential curl unzip
Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [55.4 kB]
Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
Get:4 http://deb.debian.org/debian bookworm/main amd64 Packages [8787 kB]
Get:5 http://deb.debian.org/debian bookworm-updates/main amd64 Packages [2468 B]
Get:6 http://deb.debian.org/debian-security bookworm-security/main amd64 Packages [188 kB]
Fetched 9232 kB in 2s (4113 kB/s)
Reading package lists... Done
Reading package lists... Done
...

root@02a62aa013df:/# install -m 0755 -d /etc/apt/keyrings

root@02a62aa013df:/# curl -fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc

root@02a62aa013df:/# echo "deb [arch="$(dpkg --print-architecture)" signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/debian "$(. /etc/os-release && echo "$VERSION_CODENAME")" stable" | tee /etc/apt/sources.list.d/docker.list
deb [arch=amd64 signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/debian bookworm stable

root@02a62aa013df:/# apt-get update
Hit:1 http://deb.debian.org/debian bookworm InRelease
Hit:2 http://deb.debian.org/debian bookworm-updates InRelease
Hit:3 http://deb.debian.org/debian-security bookworm-security InRelease
Get:4 https://download.docker.com/linux/debian bookworm InRelease [43.3 kB]
Get:5 https://download.docker.com/linux/debian bookworm/stable amd64 Packages [31.3 kB]
Fetched 74.6 kB in 1s (64.9 kB/s)
fatal error in libgcrypt, file ../../src/misc.c, line 92, function _gcry_fatal_error: requested algo not in md context

Fatal error: requested algo not in md context
Aborted
```

dsotirho-ucsc committed:
Disable FIPS mode in Docker container during apt-get update

Resolves error with libgcrypt on bookworm:

Fatal error in libgcrypt, requested algo not in md context

dsotirho-ucsc commented:
Successful test mounting to `/proc/sys/crypto/fips_enabled` (run on gitlab dev):

```
sudo /usr/bin/docker run \
-it \
--rm \
--network gitlab-runner-net \
--env DOCKER_HOST=tcp://gitlab-dind:2375 \
--dns 172.21.0.2 \
--dns 169.254.169.253 \
122796619775.dkr.ecr.us-east-1.amazonaws.com/docker.io/library/docker:27.3.1 \
docker run \
-it \
--volume /var/run/docker.sock:/var/run/docker.sock \
docker.gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/runner:latest \
docker run \
-it \
--volume /var/run/docker.sock:/var/run/docker.sock \
--env DOCKER_BUILDKIT=1 docker.gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/runner:latest \
bash -c 'echo 0 > /root/zero; \
printf "FROM docker.io/library/python:3.12.6-slim-bookworm\n\
RUN --mount=type=bind,source=zero,target=/proc/sys/crypto/fips_enabled cat /proc/sys/crypto/fips_enabled\n\
RUN cat /proc/sys/crypto/fips_enabled\n" > /root/Dockerfile; \
docker build --no-cache --progress=plain /root'
```

Output:
```
...
#6 [stage-0 2/3] RUN --mount=type=bind,source=zero,target=/proc/sys/crypto/fips_enabled cat /proc/sys/crypto/fips_enabled
#6 0.523 0
#6 DONE 0.7s
...
```

hannes-ucsc commented:
Simplified version below. 
- removed `DOCKER_BUILDKIT=1` since that is the default anyways
- converted the `bash -c` to just `bash` with a heredoc
- converted the `printf` for the Dockerfile to a heredoc
- removed the intermediate `docker run` of the runner image  (I don't think it reflects reality)
- ~removed the second RUN instruction (the one without the mount) from the Dockerfile~

```
sudo /usr/bin/docker run \
    --interactive \
    --rm \
    --network gitlab-runner-net \
    --env DOCKER_HOST=tcp://gitlab-dind:2375 \
    --dns 172.21.0.2 \
    --dns 169.254.169.253 \
    122796619775.dkr.ecr.us-east-1.amazonaws.com/docker.io/library/docker:27.3.1 \
    docker run \
        --interactive \
        --rm \
        --volume /var/run/docker.sock:/var/run/docker.sock \
        docker.gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/runner:latest \
        bash <<EOF1
            echo 0 > zero
            docker build --no-cache --progress=plain -f - . <<EOF2
                FROM docker.io/library/python:3.12.6-slim-bookworm
                RUN cat /proc/sys/crypto/fips_enabled
                RUN --mount=type=bind,source=zero,target=/proc/sys/crypto/fips_enabled cat /proc/sys/crypto/fips_enabled
EOF2
EOF1
```

hannes-ucsc commented:
Here's a simplified version of the reproduction as a paste-able one-liner for GitLab `dev`:

```
sudo /usr/bin/docker run \
    --interactive \
    --rm \
    --network gitlab-runner-net \
    --env DOCKER_HOST=tcp://gitlab-dind:2375 \
    --dns 172.21.0.2 \
    --dns 169.254.169.253 \
    122796619775.dkr.ecr.us-east-1.amazonaws.com/docker.io/library/docker:27.3.1 \
        docker run \
            --interactive \
            --rm \
            --volume /var/run/docker.sock:/var/run/docker.sock \
            docker.gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/runner:latest \
                docker run \
                    --interactive \
                    docker.io/library/python:3.12.6-slim-bookworm \
                    bash <<'EOF'
                        set -ex
                        apt-get update && apt-get -y install curl
                        install -m 0755 -d /etc/apt/keyrings
                        curl -fsSL https://download.docker.com/linux/debian/gpg -o /etc/apt/keyrings/docker.asc
                        echo "deb [arch="$(dpkg --print-architecture)" signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/debian "$(. /etc/os-release && echo "$VERSION_CODENAME")" stable" | tee /etc/apt/sources.list.d/docker.list
                        apt-get update
EOF
```

This ends in 

```
+ apt-get update
Hit:1 http://deb.debian.org/debian bookworm InRelease
Hit:2 http://deb.debian.org/debian bookworm-updates InRelease
Hit:3 http://deb.debian.org/debian-security bookworm-security InRelease
Get:4 https://download.docker.com/linux/debian bookworm InRelease [43.3 kB]
Get:5 https://download.docker.com/linux/debian bookworm/stable amd64 Packages [31.3 kB]
Fetched 74.6 kB in 1s (87.5 kB/s)
Reading package lists...fatal error in libgcrypt, file ../../src/misc.c, line 92, function _gcry_fatal_error: requested algo not in md context

Fatal error: requested algo not in md context
bash: line 6:   274 Aborted                 apt-get update
```

The reason is that the [InRelease](https://download.docker.com/linux/debian/dists/bookworm/InRelease) file from  Docker's package repository contains MD5 hashes of the files in the release. It also includes a SHA1 and SHA256 hash for each file. The file is signed so when `apt-get` tries to verify the signature, it must accumulate all the hashes, including the MD5s into the overall digest, using libgcrypt20 for that. Debian bookworm updated libgcrypt20. The previous version had an exception for MD5, in FIPS mode. The new version does not. A [bug report](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1014517) was filed against Debian.

Most `InRelease` files from upstream don't contain MD5 hashes anymore and that's why the first `apt-get update` in the above reproduction succeeds.

hannes-ucsc commented:
As to why the `RUN --mount` work-around fails with Docker Desktop on macOS: The host's `/proc` FS is mounted in all containers. This is the case even with our DinD setup on the EC2 instance with Amazon Linux, since all containers ultimately share the kernel, and therefore the /proc filesystem, with the host, even containers managed by the DinD daemon rather than the hosts Docker daemon.

With Docker Desktop with macOS the host is not the physical machine but a VM running a minimal Linux distro (LinuxKit). This distro and kernel don't provide `/proc/sys/crypto` and faking it is not easy. Even if we got it to work, `runc` would then refuse the mount. It only allows certain bind mounts in `/proc`, like `/proc/sys/net`. I am unsure as to why the DinD `runc` on the instance allows it but the workaround proves that it does.

The key observation is that we don't need the `RUN --mount` workaround when building the Dockerfile on macOS. The problem is not reproducible there because `/proc/sys/crypto` is completely absent and libgcrypt therefore doesn't attempt to enforce any FIPS restrictions.

We can 

```diff
Index: Dockerfile
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Dockerfile b/Dockerfile
--- a/Dockerfile	(revision 70f573de7639abf1180a0d78b575ea7b71a7a64a)
+++ b/Dockerfile	(date 1730396147875)
@@ -46,7 +46,8 @@
 RUN install -m 0755 -d /etc/apt/keyrings
 COPY --chmod=0644 bin/keys/docker-apt-keyring.pgp /etc/apt/keyrings/docker.gpg
 ARG azul_docker_version
-RUN --mount=type=bind,source=fips_enabled,target=/proc/sys/crypto/fips_enabled \
+ARG azul_proc_sys_crypto=/tmp
+RUN --mount=type=bind,source=fips_enabled,target=${azul_proc_sys_crypto}/fips_enabled \
     set -o pipefail \
     && ( \
       echo "deb [arch="$(dpkg --print-architecture)" signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/debian "$(. /etc/os-release && echo "$VERSION_CODENAME")" stable" \
```

and then override `azul_proc_sys_crypto` on the EC2 instance to actually point to `/proc/sys/crypto`.

Alternatively, we could disable FIPS mode on the instance.

Independently, we should ask Docker to remove the MD5 hashes from their `InRelease` file for bookworm. Either workaround (`RUN --mount … target=${azul_proc_sys_crypto}/fips_enabled …` or the FIPS mode disablement) could be reverted when they do.

Assignee to try the first workaround on `tempdev`. I verified it on my local Mac already.

hannes-ucsc commented:
> Independently, we should ask Docker to remove the MD5 hashes from their InRelease file for bookworm.

https://github.com/docker/docker-ce-packaging/issues/1096

I also created https://github.com/DataBiosphere/azul/issues/6675 so that we can add a FIXME referring to it in this PR.

dsotirho-ucsc committed:
Add FIXME (#6675)

dsotirho-ucsc commented:
Successfully build on GitLab tempdev

```
[0KRunning with gitlab-runner 17.4.1 (32fe5502)[0;m
[0K  on Azul #2 ckYQpq9g, system ID: r_Id9pnxxRY3e8[0;m
section_start:1731013406:prepare_executor
[0K[0K[36;1mPreparing the "docker" executor[0;m[0;m
[0KUsing Docker executor with image docker.gitlab.temp.gi.ucsc.edu/ucsc/azul/runner:latest ...[0;m

...

#12 [stage-0  8/12] RUN --mount=type=bind,source=fips_enabled,target=/proc/sys/crypto/fips_enabled     set -o pipefail     && (       echo "deb [arch="$(dpkg --print-architecture)" signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/debian "$(. /etc/os-release && echo "$VERSION_CODENAME")" stable"       | tee /etc/apt/sources.list.d/docker.list     )     && apt-get update     && version=$(apt-cache madison docker-ce | awk '{ print $3 }' | grep -P "^5:\Q27.3.1\E" | head -1)     && test -n "$version"     && apt-get -y install docker-ce=$version docker-ce-cli=$version docker-buildx-plugin
#12 0.458 deb [arch=amd64 signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/debian bookworm stable
#12 0.549 Hit:1 http://deb.debian.org/debian bookworm InRelease
#12 0.550 Hit:2 http://deb.debian.org/debian bookworm-updates InRelease
#12 0.550 Hit:3 http://deb.debian.org/debian-security bookworm-security InRelease
#12 0.560 Get:4 https://download.docker.com/linux/debian bookworm InRelease [43.3 kB]
#12 0.952 Get:5 https://download.docker.com/linux/debian bookworm/stable amd64 Packages [31.3 kB]
#12 0.988 Fetched 74.6 kB in 0s (157 kB/s)
#12 0.988 Reading package lists...
#12 2.911 Reading package lists...
#12 3.871 Building dependency tree...
#12 4.142 Reading state information...
#12 4.578 The following additional packages will be installed:
#12 4.578   apparmor containerd.io dbus dbus-bin dbus-daemon dbus-session-bus-common
#12 4.578   dbus-system-bus-common dbus-user-session dmsetup docker-ce-rootless-extras
#12 4.579   docker-compose-plugin git git-man iptables less libapparmor1 libargon2-1
#12 4.579   libcbor0.8 libcryptsetup12 libcurl3-gnutls libdbus-1-3 libdevmapper1.02.1
#12 4.580   libedit2 liberror-perl libfdisk1 libfido2-1 libglib2.0-0 libglib2.0-data
#12 4.580   libicu72 libip4tc2 libip6tc2 libjson-c5 libkmod2 libltdl7 libmnl0
#12 4.580   libnetfilter-conntrack3 libnfnetlink0 libnftables1 libnftnl11 libnss-systemd
#12 4.582   libpam-systemd libproc2-0 libslirp0 libsystemd-shared libxext6 libxml2
#12 4.584   libxmuu1 libxtables12 nftables openssh-client pigz procps psmisc
#12 4.584   shared-mime-info slirp4netns systemd systemd-sysv systemd-timesyncd xauth
#12 4.584   xdg-user-dirs
#12 4.589 Suggested packages:
#12 4.589   apparmor-profiles-extra apparmor-utils aufs-tools cgroupfs-mount
#12 4.589   | cgroup-lite gettext-base git-daemon-run | git-daemon-sysvinit git-doc
#12 4.589   git-email git-gui gitk gitweb git-cvs git-mediawiki git-svn firewalld kmod
#12 4.589   low-memory-monitor keychain libpam-ssh monkeysphere ssh-askpass
#12 4.589   systemd-container systemd-homed systemd-userdbd systemd-boot
#12 4.589   systemd-resolved libqrencode4 libtss2-esys-3.0.2-0 libtss2-mu0 libtss2-rc0
#12 4.589   polkitd | policykit-1
#12 4.980 The following NEW packages will be installed:
#12 4.980   apparmor containerd.io dbus dbus-bin dbus-daemon dbus-session-bus-common
#12 4.980   dbus-system-bus-common dbus-user-session dmsetup docker-buildx-plugin
#12 4.980   docker-ce docker-ce-cli docker-ce-rootless-extras docker-compose-plugin git
#12 4.980   git-man iptables less libapparmor1 libargon2-1 libcbor0.8 libcryptsetup12
#12 4.980   libcurl3-gnutls libdbus-1-3 libdevmapper1.02.1 libedit2 liberror-perl
#12 4.981   libfdisk1 libfido2-1 libglib2.0-0 libglib2.0-data libicu72 libip4tc2
#12 4.981   libip6tc2 libjson-c5 libkmod2 libltdl7 libmnl0 libnetfilter-conntrack3
#12 4.981   libnfnetlink0 libnftables1 libnftnl11 libnss-systemd libpam-systemd
#12 4.982   libproc2-0 libslirp0 libsystemd-shared libxext6 libxml2 libxmuu1
#12 4.983   libxtables12 nftables openssh-client pigz procps psmisc shared-mime-info
#12 4.983   slirp4netns systemd systemd-sysv systemd-timesyncd xauth xdg-user-dirs
#12 5.035 0 upgraded, 63 newly installed, 0 to remove and 0 not upgraded.
#12 5.035 Need to get 157 MB of archives.
#12 5.035 After this operation, 587 MB of additional disk space will be used.
#12 5.035 Get:1 http://deb.debian.org/debian bookworm/main amd64 libargon2-1 amd64 0~20171227-0.3+deb12u1 [19.4 kB]
#12 5.039 Get:2 http://deb.debian.org/debian bookworm/main amd64 dmsetup amd64 2:1.02.185-2 [82.0 kB]

...

#12 30.33 Created symlink /etc/systemd/system/multi-user.target.wants/docker.service → /lib/systemd/system/docker.service.

#12 30.53 Created symlink /etc/systemd/system/sockets.target.wants/docker.socket → /lib/systemd/system/docker.socket.

#12 30.54 Setting up docker-ce-rootless-extras (5:27.3.1-1~debian.12~bookworm) ...
#12 30.55 Processing triggers for libc-bin (2.36-9+deb12u8) ...
#12 DONE 30.8s

...

 [33m2 warnings found (use --debug to expand):
[0m - InvalidDefaultArgInFrom: Default value for ARG ${azul_docker_registry}${azul_python_image} results in empty or invalid base image name (line 3)
 - RedundantTargetPlatform: Setting platform to predefined ${TARGETPLATFORM} in FROM is redundant as this is the default behavior (line 3)
rm fips_enabled
docker push docker.gitlab.temp.gi.ucsc.edu/ucsc/azul/dev:20
The push refers to repository [docker.gitlab.temp.gi.ucsc.edu/ucsc/azul/dev]
78413fa7c68a: Preparing
56ef5562ead4: Preparing
5f70bf18a086: Preparing
ed68b401b8fa: Preparing
4461181adc38: Preparing
8b9106654bd8: Preparing
79618bf9b484: Preparing
dc7acb92fba9: Preparing
46f9f55e0ca9: Preparing
2e71ef0985a5: Preparing
e2e1ce994dd4: Preparing
5c5b1cd04947: Preparing
ecfc14c16318: Preparing
de732417db5d: Preparing
8e2ab394fabf: Preparing
46f9f55e0ca9: Waiting
2e71ef0985a5: Waiting
e2e1ce994dd4: Waiting
5c5b1cd04947: Waiting
ecfc14c16318: Waiting
de732417db5d: Waiting
8e2ab394fabf: Waiting
8b9106654bd8: Waiting
79618bf9b484: Waiting
dc7acb92fba9: Waiting
5f70bf18a086: Layer already exists
ed68b401b8fa: Pushed
56ef5562ead4: Pushed
8b9106654bd8: Pushed
79618bf9b484: Pushed
46f9f55e0ca9: Pushed
2e71ef0985a5: Pushed
5c5b1cd04947: Pushed
ecfc14c16318: Pushed
de732417db5d: Pushed
dc7acb92fba9: Pushed
8e2ab394fabf: Pushed
e2e1ce994dd4: Pushed
78413fa7c68a: Pushed
4461181adc38: Pushed
20: digest: sha256:68a26d2a61fbcb658ad755b9604c394da6f4c924d1a00a371f0ba9b141dee868 size: 3464
section_end:1731013631:step_script
[0Ksection_start:1731013631:cleanup_file_variables
[0K[0K[36;1mCleaning up project directory and file based variables[0;m[0;m
section_end:1731013632:cleanup_file_variables
[0K[32;1mJob succeeded[0;m
```

dsotirho-ucsc commented:
[5951_IT_2024-11-12.txt](https://github.com/user-attachments/files/17724454/5951_IT_2024-11-12.txt)

dsotirho-ucsc commented:
This PR was previously approved however failed in sandbox (see [comment](https://github.com/DataBiosphere/azul/pull/5951#issuecomment-2400943966)). The latest version of this PR was successfully tested in GitLab `tempdev`.

achave11-ucsc reviewed (changes requested):
Consider upgrading to version [3.12.7](https://www.python.org/downloads/release/python-3127/).

dsotirho-ucsc commented:
[5951_IT_2024-11-13.txt](https://github.com/user-attachments/files/17740216/5951_IT_2024-11-13.txt)

achave11-ucsc commented:
@dsotirho-ucsc , question, if the commit titled 'Implement referencing.Registry in JSON schema validation (https://github.com/DataBiosphere/azul/issues/6666)' resolves the referenced issue, should the deprecation warning (i.e. the FIXME site) also be removed? Or are there other tests that raise the deprecation warning that wouldn't be addressed by the changes in said commit (IIRC various test cases were affected by this warning) ?

achave11-ucsc reviewed (dismissed):
Excellent work! ✅

hannes-ucsc commented on Makefile:
```diff
@@ -36,15 +36,18 @@ $(eval $(call requirements,_runtime_deps,requirements_pip,,))
 	       --build-arg azul_terraform_version=$$(azul_terraform_version) \
 	       --build-arg PIP_DISABLE_PIP_VERSION_CHECK=$$(PIP_DISABLE_PIP_VERSION_CHECK) \
 	       --build-arg make_target=requirements$2 \
+	       --build-arg azul_proc_sys_crypto=$$(azul_proc_sys_crypto) \
```
Environment variables need to be documented in `environment.py` at the project root …

hannes-ucsc commented on Dockerfile:
```diff
@@ -46,7 +46,11 @@ RUN mkdir terraform \
-RUN set -o pipefail \
+# FIXME: Remove mounting of fips_enabled
+#        https://github.com/DataBiosphere/azul/issues/6675
+ARG azul_proc_sys_crypto=/tmp
```
… and that's also where the default value is specified.

dsotirho-ucsc commented:
[5951_IT_2024-11-14.txt](https://github.com/user-attachments/files/17760252/5951_IT_2024-11-14.txt)

dsotirho-ucsc commented:
[5951_IT_2024-11-19.txt](https://github.com/user-attachments/files/17819659/5951_IT_2024-11-19.txt)

hannes-ucsc commented on environment.py:
```diff
@@ -901,5 +901,15 @@ def env() -> Mapping[str, Optional[str]]:
-        'azul_vpn_subnet': None
+        'azul_vpn_subnet': None,
+
+        # FIXME: Remove azul_proc_sys_crypto
+        #        https://github.com/DataBiosphere/azul/issues/6675
+        # When building the Azul dockerfile on a FIPS mode enabled system (e.g.
+        # GitLab), this variable should be set to `/proc/sys/crypto`, the path
+        # where a `fips_enabled` file will be mounted. This is required for the
+        # command `apt-get update` to succeed which would otherwise fail on
+        # Debian bookworm with FIPS mode enabled.
+        #
```
Empty lines between paragraphs.

```suggestion
        # When building the Azul dockerfile on a FIPS mode enabled system (e.g.
        # GitLab), this variable should be set to `/proc/sys/crypto`, the path
        # where a `fips_enabled` file will be mounted. This is required for the
        # command `apt-get update` to succeed which would otherwise fail on
        # Debian bookworm with FIPS mode enabled.
        #
        # FIXME: Remove azul_proc_sys_crypto
        #        https://github.com/DataBiosphere/azul/issues/6675
        #
```

hannes-ucsc reviewed (changes requested):
This team uses empty lines separating logical paragraphs in markdown, comments and commit messages, as do most people. The commits that have a commit message body should have an empty line separating the commit message title.

dsotirho-ucsc commented:
[5951_IT_2024-11-20.txt](https://github.com/user-attachments/files/17837888/5951_IT_2024-11-20.txt)

hannes-ucsc commented on scripts/sell_unused_slots.py:
```diff
@@ -76,6 +76,9 @@ def _list_contribution_lambda_functions(cls) -> list[Lambda]:
     def _lambda_invocation_counts(self) -> dict[Lambda, int]:
+        # FIXME: DeprecationWarning for datetime methods in Python 3.12
+        #        https://github.com/DataBiosphere/azul/issues/5953
+        #
```
```suggestion
```

hannes-ucsc commented on src/azul/dss.py:
```diff
@@ -9,6 +9,9 @@
 def new_version():
+    # FIXME: DeprecationWarning for datetime methods in Python 3.12
+    #        https://github.com/DataBiosphere/azul/issues/5953
+    #
```
Same.

hannes-ucsc commented on test/azul_test_case.py:
```diff
@@ -147,6 +147,12 @@ class RE(str):
+
+                # FIXME: DeprecationWarning for datetime methods in Python 3.12
+                #        https://github.com/DataBiosphere/azul/issues/5953
+                #
```
Same.

hannes-ucsc commented on test/service/test_manifest.py:
```diff
@@ -1132,6 +1132,9 @@ def test_bdbag_manifest_remove_redundant_entries(self):
         """
+        # FIXME: DeprecationWarning for datetime methods in Python 3.12
+        #        https://github.com/DataBiosphere/azul/issues/5953
+        #
```
Same.

dsotirho-ucsc commented:
[5951_IT_2024-11-22.txt](https://github.com/user-attachments/files/17873729/5951_IT_2024-11-22.txt)

hannes-ucsc commented on src/humancellatlas/data/metadata/helpers/schema_validation.py:
```diff
@@ -25,18 +32,28 @@ class SchemaValidator:
+    @cached_property
+    def validator(self) -> Draft202012Validator:
+        registry = Registry(retrieve=self._retrieve_resource)
+        return Draft202012Validator(schema={}, registry=registry, format_checker=FormatChecker())
```
L

Please coordinate with @achave11-ucsc as he may have a PR that adds a warning suppression and FIXME for this.

hannes-ucsc commented on .gitlab-ci.yml:
```diff
@@ -11,6 +11,7 @@ workflow:
 variables:
   azul_image: $CI_REGISTRY_IMAGE
   azul_image_tag: $CI_PIPELINE_ID
+  azul_proc_sys_crypto: /proc/sys/crypto
```
FIXME comment here, too.

hannes-ucsc commented on Makefile:
```diff
@@ -36,15 +36,18 @@ $(eval $(call requirements,_runtime_deps,requirements_pip,,))
+	       --build-arg azul_proc_sys_crypto=$$(azul_proc_sys_crypto) \
 	       --tag $$(azul_image)$3:$$(azul_image_tag) \
 	       .
+	rm fips_enabled
```
Since this isn't going to be invoked if the preceding command fail, we should probably not have it at all.

But double-check that the `clean` target removes this file.

dsotirho-ucsc commented:
[5951_IT_2024-11-26.txt](https://github.com/user-attachments/files/17927507/5951_IT_2024-11-26.txt)

hannes-ucsc reviewed (changes requested):
Conflicts.

dsotirho-ucsc commented:
[5951_IT_2024-12-02.txt](https://github.com/user-attachments/files/17987245/5951_IT_2024-12-02.txt)

hannes-ucsc commented on .gitignore:
```diff
@@ -5,6 +5,9 @@ __pycache__/
 /fedramp_inventory.xlsx
+# FIXME: Remove fips_enabled
+#        https://github.com/DataBiosphere/azul/issues/6675
+fips_enabled
```
```suggestion
/fips_enabled
```

hannes-ucsc commented on Makefile:
```diff
@@ -36,13 +36,17 @@ $(eval $(call requirements,_runtime_deps,requirements_pip,,))
 	       --build-arg azul_terraform_version=$$(azul_terraform_version) \
 	       --build-arg PIP_DISABLE_PIP_VERSION_CHECK=$$(PIP_DISABLE_PIP_VERSION_CHECK) \
 	       --build-arg make_target=requirements$2 \
+	       --build-arg azul_proc_sys_crypto=$$(azul_proc_sys_crypto) \
```
```suggestion
	       --build-arg azul_docker_registry=$$(azul_docker_registry) \
	       --build-arg azul_python_image=$$(azul_python_image) \
	       --build-arg azul_docker_version=$$(azul_docker_version) \
	       --build-arg azul_terraform_version=$$(azul_terraform_version) \
	       --build-arg azul_proc_sys_crypto=$$(azul_proc_sys_crypto) \
	       --build-arg PIP_DISABLE_PIP_VERSION_CHECK=$$(PIP_DISABLE_PIP_VERSION_CHECK) \
	       --build-arg make_target=requirements$2 \
```

hannes-ucsc commented on environment.py:
```diff
@@ -901,5 +901,16 @@ def env() -> Mapping[str, Optional[str]]:
-        'azul_vpn_subnet': None
+        'azul_vpn_subnet': None,
+
+        # When building the Azul dockerfile on a FIPS mode enabled system (e.g.
```
```suggestion
        # When building the Azul image on a FIPS mode enabled system (e.g.
```

hannes-ucsc commented on environment.py:
```diff
@@ -901,5 +901,16 @@ def env() -> Mapping[str, Optional[str]]:
+        # FIXME: Remove azul_proc_sys_crypto
+        #        https://github.com/DataBiosphere/azul/issues/6675
+        #
+        'azul_proc_sys_crypto': '/tmp'
```
This should go after the definition of `azul_terraform_version`.

dsotirho-ucsc commented:
[5951_IT_2024-12-04.txt](https://github.com/user-attachments/files/18011919/5951_IT_2024-12-04.txt)

hannes-ucsc commented:
### Security design review

- [x] Security design review completed; this PR does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with Terra 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

dsotirho-ucsc committed:
[R] Ran `make requirements_update`
